src/Razor/src/Microsoft.AspNetCore.Razor.ExternalAccess.RoslynWorkspace/RazorWorkspaceListenerBase.cs

```diff
@@ -0,0 +1,246 @@
+        if (_stream is null ||
+            project is not
+            {
+                Language: LanguageNames.CSharp
+            })
+        {
+            return;
+        }
+
+        // Don't queue work for projects that don't have a dynamic file
+        if (!_projectsWithDynamicFile.TryGetValue(project.Id, out var _))
+        {
+            return;
+        }
+
+        var projectId = project.Id;
+        _workQueue.AddWork(new UpdateWork(projectId));
+    }
+
+    private void RemoveProject(Project project)
+    {
+        if (ImmutableInterlocked.TryRemove(ref _projectsWithDynamicFile, project.Id, out var _))
+        {
+            var intermediateOutputPath = Path.GetDirectoryName(project.CompilationOutputInfo.AssemblyPath);
+            if (intermediateOutputPath is null)
+            {
+                _logger?.LogTrace("intermediatePath is null, skipping notification of removal for {projectId}", project.Id);
+                return;
+            }
+
+            _workQueue.AddWork(new RemovalWork(project.Id, intermediateOutputPath));
+        }
+    }
+
+    private Task ReportRemovalAsync(RemovalWork unit, CancellationToken cancellationToken)
+    {
+        _logger?.LogTrace("Reporting removal of {projectId}", unit.ProjectId);
+        return _stream.AssumeNotNull().WriteProjectInfoRemovalAsync(unit.IntermediateOutputPath, cancellationToken);
+    }
+
+    private async ValueTask UpdateCurrentProjectsAsync(ImmutableArray<Work> work, CancellationToken cancellationToken)
+    {
+        var solution = _workspace.AssumeNotNull().CurrentSolution;
+        _stream.AssumeNotNull();
+
+        await CheckConnectionAsync(_stream, cancellationToken).ConfigureAwait(false);
+
+        foreach (var unit in work)
+        {
+            try
+            {
+                if (_disposeTokenSource.IsCancellationRequested)
+                {
+                    return;
+                }
+
+                if (unit is RemovalWork removalWork)
+                {
+                    await ReportRemovalAsync(removalWork, cancellationToken).ConfigureAwait(false);
+                }
+
+                var project = solution.GetProject(unit.ProjectId);
+                if (project is null)
+                {
+                    _logger?.LogTrace("Project {projectId} is not in workspace", unit.ProjectId);
+                    continue;
+                }
+
+                await UpdateProjectAsync(project, solution, cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                _logger?.LogError(ex, "Encountered exception while processing unit: {message}", ex.Message);
+            }
+        }
+
+        await _stream.AssumeNotNull().FlushAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    private void Workspace_WorkspaceChanged(object? sender, WorkspaceChangeEventArgs e)
+    {
+        switch (e.Kind)
+        {
+            case WorkspaceChangeKind.SolutionChanged:
```

Comment:
SolutionChanged could also theoretically have projects removed.

Reply:
hmm... we weren't handling that before. I'm guessing that was wrong. Do we need to do a diff of the ProjectIds?